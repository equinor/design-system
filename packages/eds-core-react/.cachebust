# cache bust 1761833449
