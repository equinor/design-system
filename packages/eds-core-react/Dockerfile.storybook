FROM nginx:1.26.1-alpine

WORKDIR /app

COPY ./packages/eds-core-react/storybook-build ./
COPY ./packages/eds-core-react/nginx.conf /etc/nginx/conf.d/default.conf
