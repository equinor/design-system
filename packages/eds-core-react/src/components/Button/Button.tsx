import { forwardRef, ButtonHTMLAttributes } from 'react'
import styled, { css, ThemeProvider } from 'styled-components'
import { token as buttonToken } from './tokens'
import { ButtonTokenSet, ButtonToken } from './Button.types'
import {
  typographyTemplate,
  bordersTemplate,
  outlineTemplate,
  spacingsTemplate,
  useToken,
  OverridableComponent,
} from '@equinor/eds-utils'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '../EdsProvider'

type Colors = 'primary' | 'secondary' | 'danger'
type Variants =
  | 'contained'
  | 'contained_icon'
  | 'outlined'
  | 'ghost'
  | 'ghost_icon'

const getVariant = (
  tokenSet: ButtonTokenSet,
  variant: Variants,
): ButtonToken => {
  switch (variant) {
    case 'ghost':
      return tokenSet.ghost
    case 'ghost_icon':
      return tokenSet.ghost_icon
    case 'outlined':
      return tokenSet.outlined
    case 'contained_icon':
      return tokenSet.contained_icon
    case 'contained':
    default:
      return tokenSet.contained
  }
}

const getToken = (variant: Variants, color: Colors): ButtonToken => {
  switch (color) {
    case 'danger':
      return getVariant(buttonToken.danger, variant)
    case 'secondary':
      return getVariant(buttonToken.secondary, variant)
    case 'primary':
    default:
      return getVariant(buttonToken.primary, variant)
  }
}

const Inner = styled.span`
  display: grid;
  grid-gap: var(--eds_button__gap, 8px);
  grid-auto-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;

  & > :is(svg, img) {
    margin-top: var(--eds_button__icon__margin_y, 0);
    margin-bottom: var(--eds_button__icon__margin_y, 0);
  }
`

const ButtonBase = styled.button(({ theme }: { theme: ButtonToken }) => {
  const { states, clickbound } = theme
  const { focus, hover, disabled } = states

  return css`
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    text-decoration: none;
    position: relative;
    cursor: pointer;
    display: inline-block;
    background: ${theme.background};
    height: ${theme.height};
    width: ${theme.width};

    svg {
      justify-self: center;
    }

    ${spacingsTemplate(theme.spacings)}
    ${bordersTemplate(theme.border)}
    ${typographyTemplate(theme.typography)}

    &::before {
      position: absolute;
      top: 0;
      left: 0;
      width: auto;
      min-height: auto;
      content: '';
    }

    &::after {
      position: absolute;
      top: -${clickbound?.offset?.top};
      left: -${clickbound?.offset?.left};
      width: ${clickbound?.width};
      height: ${clickbound?.height};
      content: '';
    }

    @media (hover: hover) and (pointer: fine) {
      &:hover {
        background: ${hover.background};
        color: ${hover.typography?.color};
        ${bordersTemplate(hover?.border)};
      }
    }

    &:focus {
      outline: none;
    }

    &[data-focus-visible-added]:focus {
      ${outlineTemplate(focus.outline)}
    }
    &:focus-visible {
      ${outlineTemplate(focus.outline)}
    }
    /* Get rid of ff focus border for buttons */
    &::-moz-focus-inner {
      border: 0;
    }

    &:disabled,
    &[aria-disabled='true'] {
      cursor: not-allowed;
      background: ${disabled.background};
      ${bordersTemplate(disabled.border)};
      ${typographyTemplate(disabled.typography)};

      @media (hover: hover) and (pointer: fine) {
        &:hover {
          background: ${disabled.background};
        }
      }
    }
  `
})

export type ButtonProps = {
  /**  Specifies color */
  color?: Colors
  /** Specifies which variant to use */
  variant?: Variants
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Type of button
   * @default 'button'
   */
  type?: string
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
  forwardRef(function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      ...other
    },
    ref,
  ) {
    const { density } = useEds()
    const token = useToken({ density }, getToken(variant, color))

    const as = href && !disabled ? 'a' : other.as ? other.as : 'button'

    const type = href || other.as ? undefined : 'button'

    tabIndex = disabled ? -1 : tabIndex

    const buttonProps = {
      ref,
      as,
      href,
      type,
      disabled,
      tabIndex,
      ...other,
    }

    return (
      <ThemeProvider theme={token}>
        <ButtonBase {...buttonProps}>
          {fullWidth ? (
            <InnerFullWidth>{children}</InnerFullWidth>
          ) : (
            <Inner>{children}</Inner>
          )}
        </ButtonBase>
      </ThemeProvider>
    )
  })
