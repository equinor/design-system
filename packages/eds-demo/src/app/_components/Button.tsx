import React from "react";

type ButtonProps = {
  className?: string;
  icon?: React.ReactNode;
  onClick?: () => void;
  children?: React.ReactNode;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

export const PrimaryButton = ({
  className,
  icon,
  onClick,
  children = "Save",
  ...rest
}: ButtonProps) => {
  return (
    <button
      data-debug
      type="button"
      className={`select-none h-[36px] px-4 py-2.5 rounded border-opacity-0 justify-start gap-2 flex items-center button button--primary ${
        className || ""
      }`}
      onClick={onClick}
      {...rest}
    >
      {icon}
      <div className="text-sm leading-none">{children}</div>
    </button>
  );
};

export const SecondaryButton = ({
  className,
  icon,
  onClick,
  children = "Cancel",
  ...rest
}: ButtonProps) => {
  return (
    <button
      data-debug
      type="button"
      className={`select-none h-[36px] px-4 py-2.5 bg-opacity-0 rounded border justify-start gap-2 flex items-center ${
        className || ""
      }`}
      onClick={onClick}
      {...rest}
    >
      {icon}
      <div className="text-sm leading-none">{children}</div>
    </button>
  );
};
