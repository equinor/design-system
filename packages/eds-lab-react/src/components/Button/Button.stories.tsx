import { useState, useEffect } from 'react'
import {
  Button,
  Icon,
  ButtonProps,
  EdsProvider,
  Density,
} from '@equinor/eds-core-react'
import styled, { css, FlattenSimpleInterpolation } from 'styled-components'
import { StoryFn, Meta } from '@storybook/react'

const Wrapper = styled.div<{ css: FlattenSimpleInterpolation }>`
  margin: 32px;
  display: grid;
  grid-gap: 32px;
  grid-template-columns: repeat(4, fit-content(100%));
  align-items: start;
`

const meta: Meta<typeof Button> = {
  title:
    'Core-react experimental features/css-variables/Button With Relative Units',
  component: Button,
  parameters: {
    docs: {
      description: {
        component: `Overriding exposed css variables.
        `,
      },
    },
  },
}

export default meta

export const Default: StoryFn<ButtonProps> = () => {
  const [density, setDensity] = useState<Density>('comfortable')

  useEffect(() => {
    // Simulate user change
    setDensity('compact')
  }, [density])

  const vars = css`
    --eds_button__font_size: calc(14 / 16 * 1rem);
    --eds_button__radius: calc(4 / 14 * 1em);
    --eds_button__height: auto;
    --eds_button__height_compact: auto;
    --eds_button__gap: calc(8 / 14 * 1em);
    --eds_button__border_width: 1px;
    --eds_button__padding_y: calc(
      10 / 14 * 1em - var(--eds_button__border_width)
    );
    --eds_button__padding_y_compact: calc(
      4 / 14 * 1em - var(--eds_button__border_width)
    );
    --eds_button__padding_x: calc(16 / 14 * 1em);
    --eds_button__icon__size: 24px;
    --eds_button__icon__margin_y: -4px; // icon = 24px, line-height: 16px, (24 - 16) / 2 = 4px
    --eds_button__fullwidth__icon__margin_x: calc(8 / 14 * 1em);
    --eds_button__fullwidth__label__margin_x: calc(32 / 14 * 1em);
  `

  return (
    <Wrapper css={vars}>
      <Button variant="outlined">Outlined</Button>
      <Button>
        Contained
        <br />
        multiline
      </Button>
      <EdsProvider density={density}>
        <Button variant="outlined">Compact</Button>
        <Button variant="outlined">
          Compact
          <br />
          multiline
        </Button>
      </EdsProvider>
      <Button>
        <Icon name="save" title="save" />
        Icon
      </Button>
      <Button>
        <Icon name="save" title="save" />
        Icon
        <br />
        multiline
      </Button>
    </Wrapper>
  )
}
