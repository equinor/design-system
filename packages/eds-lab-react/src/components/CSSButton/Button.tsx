import {
  forwardRef,
  ElementType,
  ButtonHTMLAttributes,
  createElement,
} from 'react'
import { clsx } from 'clsx'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '@equinor/eds-core-react'
import './button.css'
//** Injecting root css properties needs to be solved */
import '@equinor/eds-tokens/tokens.css'

export type ButtonProps = {
  /**  Specifies color */
  color?: 'primary' | 'secondary' | 'danger'
  /** Specifies which variant to use */
  variant?: 'contained' | 'outlined' | 'ghost' | 'ghost_icon'
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Change html element. */
  as?: ElementType
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const CSSButton = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      className,
      ...other
    },
    ref,
  ) {
    const { density } = useEds()

    const as: ElementType =
      href && !disabled ? 'a' : other.as ? other.as : 'button'
    // const type = href || other.as ? undefined : 'button'

    const classNames: string = clsx([
      'eds-typography-navigation__btn',
      'eds-btn',
      color,
      variant,
      density,
      className,
    ])

    const buttonProps = {
      ref,
      href,
      disabled,
      tabIndex,
      className: classNames,
      ...other,
    }

    return createElement(
      as,
      buttonProps,
      fullWidth ? (
        <InnerFullWidth>{children}</InnerFullWidth>
      ) : (
        <span className="eds-btn-inner">{children}</span>
      ),
    )
  },
)
