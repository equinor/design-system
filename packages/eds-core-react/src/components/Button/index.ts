import { Button as ButtonWrapper, ButtonProps } from './Button'
import { ButtonGroup, ButtonGroupProps } from './ButtonGroup/ButtonGroup'
import { ToggleButton, ToggleButtonProps } from './ToggleButton/ToggleButton'

type ButtonCompoundProps = typeof ButtonWrapper & {
  Group: typeof ButtonGroup
  Toggle: typeof ToggleButton
}

const Button = ButtonWrapper as ButtonCompoundProps
Button.Group = ButtonGroup
Button.Toggle = ToggleButton

Button.Group.displayName = 'Button.Group'
Button.Toggle.displayName = 'Button.Toggle'

export { Button }
export type { ButtonProps, ButtonGroupProps, ToggleButtonProps }
