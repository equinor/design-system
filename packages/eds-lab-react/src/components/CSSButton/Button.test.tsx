/* eslint-disable no-undef */
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import styled from 'styled-components'
import { Icon } from '@equinor/eds-core-react'
import { save } from '@equinor/eds-icons'
import { CSSButton } from './Button'
import React from 'react'

Icon.add({ save })

const StyledButton = styled(CSSButton)`
  position: relative;
  height: 100px;
  width: 100px;
`

const MarginButton = styled(CSSButton)`
  margin: 12px;
`

describe('Button', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<CSSButton>Button</CSSButton>)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Has provided icon when variant is icon', () => {
    render(
      <CSSButton variant="ghost_icon">
        <Icon name="save" title="save me test" />
      </CSSButton>,
    )
    expect(screen.getByLabelText('save me test')).toBeInTheDocument()
    expect(screen.queryByTestId('eds-icon-path')).toHaveAttribute(
      'd',
      save.svgPathData,
    )
  })
  it('Can extend the css for the component', () => {
    render(<StyledButton>Test me!</StyledButton>)
    const button = screen.getByRole('button')

    expect(button).toHaveStyleRule('position', 'relative')
    expect(button).toHaveStyleRule('height', '100px')
    expect(button).toHaveStyleRule('width', '100px')
  })
  it('onSubmit is called one time when used in form', () => {
    const onSubmit = jest.fn((e: React.FormEvent<HTMLFormElement>) =>
      e.preventDefault(),
    )

    render(
      <form onSubmit={onSubmit}>
        <CSSButton type="submit">Submit button</CSSButton>
      </form>,
    )

    const submitButton = screen.queryByText('Submit button')

    fireEvent.submit(submitButton)

    expect(onSubmit).toHaveBeenCalledTimes(1)
  })

  // TODO Why do we want this behaviour?
  // it('onSubmit is ignored in form by default', () => {
  //   const onSubmit = jest.fn((e: React.FormEvent<HTMLFormElement>) =>
  //     e.preventDefault(),
  //   )

  //   render(
  //     <form onSubmit={onSubmit}>
  //       <Button>Submit button</Button>
  //     </form>,
  //   )

  //   const submitButton = screen.queryByText('Submit button')

  //   fireEvent.click(submitButton)

  //   expect(onSubmit).toHaveBeenCalledTimes(0)
  // })
  it('Has provided icon when icon and text is defined', () => {
    render(
      <CSSButton>
        <Icon name="save" title="save"></Icon>Button
      </CSSButton>,
    )
    expect(screen.getByLabelText('save')).toBeInTheDocument()
    expect(screen.queryByTestId('eds-icon-path')).toHaveAttribute(
      'd',
      save.svgPathData,
    )
  })
  it('Renders as a link when href prop is defined', () => {
    render(<CSSButton href="/" />)
    const button = screen.getByRole('link')

    expect(button).toBeInTheDocument()
  })
  it('Can change margins', () => {
    render(<MarginButton>Test me!</MarginButton>)
    const button = screen.getByRole('button')
    expect(button).toHaveStyleRule('margin', '12px')
  })
})
