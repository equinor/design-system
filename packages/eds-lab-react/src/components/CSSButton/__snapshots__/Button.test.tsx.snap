// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="eds-typography-navigation__btn eds-btn primary contained comfortable"
    tabindex="0"
  >
    <span
      class="eds-btn-inner"
    >
      Button
    </span>
  </button>
</DocumentFragment>
`;
